kep/real-load-aware-scheduling/real-load-aware-scheduling.md

```diff
@@ -0,0 +1,540 @@
+
+<img src="images/safe-sched-graph.png" alt="safe-sched-graph" width="700" height="400"/>
+
+
+Following is the algorithm:
+
+
+**Algorithm**
+
+
+1. Get the requested resource for the pod to be scheduled as, <img src="https://render.githubusercontent.com/render/math?math=r">.
+2. Get the sliding window average (<img src="https://render.githubusercontent.com/render/math?math=M">) and standard deviation (<img src="https://render.githubusercontent.com/render/math?math=V">) of resource utilization fraction (range from 0 to 1) for all types of resources (CPU, Memory, GPU, etc.) of the current node to be scored. 
+3. Calculate the score of the current node for each type of resource: <img src="https://render.githubusercontent.com/render/math?math=S_i = M %2B r %2B V">
+4. Get a score for each type of resource and bound it to [0,1]: <img src="https://render.githubusercontent.com/render/math?math=S_i = \min(S_i, 1.0)">
+5. Calculate the node priority score per resource as: <img src="https://render.githubusercontent.com/render/math?math=U_i = (1 - S_i) \times MaxPriority ">
+6. Get the final node score as: <img src="https://render.githubusercontent.com/render/math?math=U = \min_i(U_i)">
+
+**Example**
+
+
+For example, let's say we have three nodes `N1`, `N2`, and `N3`, and the pod to be scheduled have CPU and Memory requests as 500 milicores and 1 GB. All nodes have a capacity of 4 cores and 8 GB.
+
+The pod request fraction can be computed as <img src="https://render.githubusercontent.com/render/math?math=r_{cpu} = \frac{1}{8}, r_{memory} = \frac{1}{8}">. 
+
+
+<img src="images/image6.png" alt="node-capacity" width="700" height="200"/>
+
+
+Then according to steps 2 ~ 4, the mean and standard deviation of CPU and memory fraction utilization can be computed as follows:
+
+
+<img src="images/image9.png" alt="node-mn-std-util" width="600" height="100"/>
+
+
+The score for each type of resource and each node are as follows according to step 5 ~ 6:
+
+
+<img src="images/image1.png" alt="node-mn-std-util" width="600" height="200"/>
+
+
+According to the scores we have, node `N3` will be selected. The utilization fraction of nodes before and after the placement is as follows.
+
+<img src="images/image5.png" alt="before-placement" width="600" height="150"/>
+
+<img src="images/image7.png" alt="after-placement" width="600" height="150"/>
+
+
+If we plot these in mu-sigma plots, we can see the placement automatically pushes the utilization of nodes toward the diagonal line sigma = 1 - mu.
```

Comment:
So the "1" here is the constant described above. Could you help me understand its real semantics? (as I assume this will be made configurable, so end-users would need to know how to configure it)

Reply:
@Huang-Wei, good question. The `1` here indicates `100%` of utilization. What should be configured is a coefficient \ita, which indicates \mu + \ita x \sigma < 100 % and we choose \ita = 1 here. 

Here, we are assuming that the actual usage follows Gaussian distribution and we are using the 68-96-99.5 rule to model the risk of usage exceeding the node capacity. Please refer to [68-95-99.7 rule](https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule)

So when \ita gets different values, we get different confidence of not exceeding the capacity.
- \ita = 1, we have a 68% chance that the actual usage is within the node capacity.
- \ita = 2, we have a 95% chance that the actual usage is within the node capacity.
- \ita = 3, we have a 99.7% chance that the actual usage is within the node capacity.

By default, we choose \ita = 1 as we would like to improve the overall utilization.
I will improve the description in the KEP.

Reply:
Please see https://github.com/kubernetes-sigs/scheduler-plugins/pull/61/commits/56022e76e84139ac005a15f39c9eaf90b8264fdb